File: Cap 10/10.8 - Recursividade ou Recorrencia/10.8.2 - Recursividade de Cauda/recauda.swift
/*
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.8
 * 
 * Programa ..: FUNCAO COM RECURSIVDADE DE CAUDA
 */

import Foundation

func fatorialBase(_ N: UInt16, _ P: UInt64) -> UInt64 {
    if N == 0 {
        return P
    } else {
        return fatorialBase(N - 1, UInt64(N) * P)
    }
}

func fatorial(_ N: UInt16) -> UInt64 {
    return fatorialBase(N, 1)
}

var LIMITE: UInt16 = 0

print("CALCULO DE FATORIAL")
print("RECURSIVIDADE DE CAUDA\n")

print("Qual fatorial (1-20): ", terminator: "")
if let inputLIMITE = readLine(), let limite = UInt16(inputLIMITE) {
    LIMITE = limite
}

print("Fatorial = \(fatorial(LIMITE))")

File: Cap 10/10.8 - Recursividade ou Recorrencia/10.8.2 - Recursividade de Cauda/recauda.ts
/*
 * Livro .....: Algoritmos
 *              Logica para Desenvolvimento de Programacao de
 *              Computadores
 * Autores ...: Jose Augusto N. G. Manzano
 *              Jayr Figueiredo de Oliveira
 * Editora ...: Erica
 *
 * Capitulo ..: 10 - Topico 10.8
 * 
 * Programa ..: FUNCAO COM RECURSIVDADE DE CAUDA
 */

import * as readline from 'readline-sync';

function FATORIALBASE(N: number, P: number): number {
  if (N === 0)
    return P;
  else
    return FATORIALBASE(N - 1, N * P);
}

function FATORIAL(N: number): number {
  return FATORIALBASE(N, 1);
}

let LIMITE: number;

console.log("CALCULO DE FATORIAL");
console.log("RECURSIVIDADE DE CAUDA\n");

LIMITE = parseInt(readline.question("Qual fatorial (1-21): "));

console.log("Fatorial =", FATORIAL(LIMITE));
